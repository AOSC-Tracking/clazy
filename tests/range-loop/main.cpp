#include <QtCore/QList>
#include <vector>
#include <QtCore/QMap>
using namespace std;

QList<int> getQtList()
{
    return {}; // dummy, not important
}

const QList<int> getConstQtList()
{
    return {}; // dummy, not important
}

const QList<int> & getConstRefQtList()
{
    static QList<int> foo;
    return foo;
}

void testQtContainer()
{
    QList<int> qt_container;
    for (int i : qt_container) { // Warning
    }

    const QList<int> const_qt_container;
    for (int i : const_qt_container) { // OK
    }

    for (int i : getQtList()) { // Warning
    }

    for (int i : getConstQtList()) { // OK
    }

    for (int i : getConstRefQtList()) { // OK
    }

    vector<int> stl_container;
    for (int i : stl_container) { // OK
    }
}

class A {
public:
    void foo()
    {
        for (int a : m_stlVec) {} // OK
    }

    std::vector<int> m_stlVec;
};

void testQMultiMapDetach()
{
    QMultiMap<int,int> m;
    for (int i : m) {
    }
}