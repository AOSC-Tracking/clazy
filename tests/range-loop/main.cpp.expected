range-loop/main.cpp:25:5: warning: c++11 range-loop might detach Qt container (QList) [-Wclazy-range-loop]
range-loop/main.cpp:32:5: warning: c++11 range-loop might detach Qt container (QList) [-Wclazy-range-loop]
range-loop/main.cpp:59:5: warning: c++11 range-loop might detach Qt container (QMultiMap) [-Wclazy-range-loop]
range-loop/main.cpp:98:10: warning: Missing reference in foreach with sizeof(T) = 20 bytes (struct BigTrivial) [-Wclazy-range-loop]
range-loop/main.cpp:103:10: warning: Missing reference in foreach with non trivial type (struct SmallNonTrivial) [-Wclazy-range-loop]
range-loop/main.cpp:108:10: warning: Missing reference in foreach with sizeof(T) = 20 bytes (const struct BigTrivial) [-Wclazy-range-loop]
range-loop/main.cpp:128:10: warning: Missing reference in foreach with sizeof(T) = 20 bytes (struct BigTrivial) [-Wclazy-range-loop]
